Introduce TizenViewElementary

JSUYA opened the pull request:
There are two ways to draw FlutterTizenView in Tizen. (TizenWindowElementary and TizenWindowEcore)
Both ways are drawing on Window component(surface).

We try to draw FlutterTizenView on the appropriate component
provided by Tizen UIFW and support it to be place with other components.
To do that, we created a TizenView class on the same level as TizenWindow
and abstracted common methods to TizenViewBase.
we can add FlutterTizenView to other UIFW of Tizen while adding TizenView{?}(maybe Tizen NUI).
```
 TizenViewBase -> TizenView -> TizenViewElementary
                              -> TizenView{?}
                 -> TizenWindow -> TizenWindowElementary
                                -> TizenWindowEcore
```

related PR : https://github.com/flutter-tizen/flutter-tizen/pull/373

bbrto21 commented:
In my opinion, I would recommend that you rebase this PR to the master after #291 is merged.

swift-kim commented:
The name `TizenBaseHandle` is somewhat ambiguous. We'll need to discuss about the class design and hierarchy.

bbrto21 commented on shell/platform/tizen/tizen_view_elementary.cc:
```diff
@@ -0,0 +1,317 @@
+  Ecore_Evas* ecore_evas =
+      ecore_evas_ecore_evas_get(evas_object_evas_get(elm_parent_));
+
+  elm_win_ = elm_win_get(elm_parent_);
```
`elm_win_get` is not public API. maybe you need to find some other way...

JSUYA replied:
You are right. I will fix all related to `elm_win_`.

JSUYA replied:
I first removed the unnecessary window code. please check again

JSUYA commented:
> The name `TizenBaseHandle` is somewhat ambiguous. We'll need to discuss about the class design and hierarchy.

Do you have a recommended class name and hierarchy?
I've been discussing names and hierarchy with @bbrto21.
Please let me know if there is a better way

JSUYA commented:
contribute to https://github.com/flutter-tizen/flutter-tizen/issues/351

bbrto21 reviewed (commented):
please rebase to master

JSUYA commented on shell/platform/tizen/public/flutter_tizen.h:
```diff
@@ -135,6 +145,21 @@ FLUTTER_EXPORT FlutterDesktopViewRef FlutterDesktopViewCreateFromNewWindow(
+    void* elm_parent);
+
+// Returns a handle to evas image that the FlutterView is drawn to.
+FLUTTER_EXPORT void* FlutterDesktopViewGetEvasImageHandle(
```
I changed to FlutterDesktopViewGetEvasObject()

JSUYA commented on shell/platform/tizen/tizen_view_elementary.cc:
```diff
@@ -0,0 +1,291 @@
+  }
+  if (initial_geometry_.height == 0) {
+    initial_geometry_.height = ecore_evas_height;
+  }
```
I removed screen size(ecore_evas)

JSUYA commented on shell/platform/tizen/public/flutter_tizen.h:
```diff
@@ -135,6 +145,21 @@ FLUTTER_EXPORT FlutterDesktopViewRef FlutterDesktopViewCreateFromNewWindow(
     FlutterDesktopEngineRef engine);
 
+// Creates a view that hosts and displays the given engine instance.
+FLUTTER_EXPORT FlutterDesktopViewRef FlutterDesktopViewCreateFromNewView(
```
I changed to FlutterDesktopViewCreateFromElmParent()

JSUYA commented:
>In my opinion, I would recommend that you rebase this PR to the master after https://github.com/flutter-tizen/engine/pull/291 is merged.

@bbrto21 
I rebased your #291 

>The name TizenBaseHandle is somewhat ambiguous. We'll need to discuss about the class design and hierarchy.

@swift-kim I renamed it to TizenViewBase as we discussed.
And I cleaned up the level of abstraction.
More reviews are needed. Could you please review it?

bbrto21 commented on shell/platform/tizen/BUILD.gn:
```diff
@@ -233,7 +234,7 @@ template("embedder") {
 embedder("flutter_tizen_mobile") {
   target_type = "shared_library"
-  use_evas_gl_renderer = false
+  use_evas_gl_renderer = true
```
Is this a mistake?

JSUYA replied:
oh...

bbrto21 commented on shell/platform/tizen/channels/window_channel.h:
```diff
@@ -26,7 +26,7 @@ class WindowChannel {
   std::unique_ptr<MethodChannel<EncodableValue>> channel_;
 
-  // A reference to the renderer object managed by FlutterTizenView.
+  // A reference to the renderer object managed by FlutterTizenView.s
```
Remove a typo
```suggestion
  // A reference to the renderer object managed by FlutterTizenView.
```

JSUYA replied:
oh oh...

bbrto21 commented on shell/platform/tizen/flutter_tizen_ecore.cc:
```diff
@@ -52,3 +52,23 @@ FlutterDesktopViewRef FlutterDesktopViewCreateFromNewWindow(
+                                  int32_t width,
+                                  int32_t height) {
+  // Not supported.
+}
```
Please add a newline.

bbrto21 commented on shell/platform/tizen/flutter_tizen.cc:
```diff
@@ -71,7 +71,12 @@ void FlutterDesktopEngineShutdown(FlutterDesktopEngineRef engine_ref) {
 void* FlutterDesktopPluginRegistrarGetNativeWindow(
     FlutterDesktopPluginRegistrarRef registrar) {
-  return registrar->engine->view()->window()->GetWindowHandle();
+  auto view_base = registrar->engine->view()->tizen_view();
```
I think 'view_base()' is a more appropriate name than 'tizen_view()'.
what do you think?

swift-kim replied:
`tizen_view()` makes more sense. "view_base" is ambiguous because "view" can refer to either a FlutterView or a TizenView instance according to the context.

JSUYA replied:
I changed method and member name to `tizen_view_base`. I think this is more clear.

swift-kim replied:
The `_base` suffix should not be present in the real member name. The `Base` in the class name just represents that it's an abstract type for its variants.

Example: https://github.com/flutter/engine/blob/c0d8f8404a66563dbda43fe303df998063f413be/shell/platform/windows/flutter_windows_view.h#L355

JSUYA replied:
I updated code. Thank you detail review.

bbrto21 commented on shell/platform/tizen/flutter_tizen.cc:
```diff
@@ -71,7 +71,12 @@ void FlutterDesktopEngineShutdown(FlutterDesktopEngineRef engine_ref) {
 void* FlutterDesktopPluginRegistrarGetNativeWindow(
     FlutterDesktopPluginRegistrarRef registrar) {
-  return registrar->engine->view()->window()->GetWindowHandle();
+  auto view_base = registrar->engine->view()->tizen_view();
```
In this case, we prefer explicit types instead of `auto`.
see https://github.com/flutter-tizen/engine/pull/275

JSUYA replied:
I fixed it. Thank you

bbrto21 commented on shell/platform/tizen/flutter_tizen_ecore.cc:
```diff
@@ -52,3 +52,23 @@ FlutterDesktopViewRef FlutterDesktopViewCreateFromNewWindow(
+    const FlutterDesktopViewProperties& view_properties,
+    FlutterDesktopEngineRef engine,
+    void* parent) {
+  // Not supported.
```
These are public APIs, so it would be nice to provide proper logs.
and I think that level should be warn.

JSUYA replied:
I added a FIXME comment and FT_UNIMPLEMENTED() . Please review again if this is appropriate

bbrto21 commented on shell/platform/tizen/flutter_tizen_elementary.cc:
```diff
@@ -52,3 +53,54 @@ FlutterDesktopViewRef FlutterDesktopViewCreateFromNewWindow(
+
+  std::unique_ptr<flutter::TizenViewBase> view =
+      std::make_unique<flutter::TizenViewElementary>(view_geometry,
+                                                     (Evas_Object*)(parent));
```
Please use C++ style cast.

bbrto21 commented on shell/platform/tizen/flutter_tizen_view.cc:
```diff
@@ -252,7 +260,12 @@ void FlutterTizenView::OnCommit(const std::string& str) {
+    auto window = reinterpret_cast<TizenWindow*>(tizen_view_.get());
+    OnRotate(window->GetRotation());
+  } else {
+    OnRotate(0);
```
Are there any plans to connect the orientation of the application and the orientation of the view via Embedder APIs? It doesn't seem appropriate to always leave it at 0. 
If it's unfinished feature, you can add the TODO comment or use `FT_UNIMPLEMENTED` macro.

JSUYA replied:
I needed `SendWindowMetrics(..)`. updated the code

bbrto21 commented on shell/platform/tizen/public/flutter_tizen.h:
```diff
@@ -135,6 +143,21 @@ FLUTTER_EXPORT FlutterDesktopViewRef FlutterDesktopViewCreateFromNewWindow(
     const FlutterDesktopWindowProperties& window_properties,
     FlutterDesktopEngineRef engine);
 
+// Creates a view that hosts and displays the given engine instance.
+FLUTTER_EXPORT FlutterDesktopViewRef FlutterDesktopViewCreateFromElmParent(
+    const FlutterDesktopViewProperties& view_properties,
+    FlutterDesktopEngineRef engine,
+    void* parent);
+
+// Returns a handle to evas object that the FlutterView is drawn to.
+FLUTTER_EXPORT void* FlutterDesktopViewGetEvasObject(
+    FlutterDesktopViewRef view);
+
+// Resize the FlutterView.
+FLUTTER_EXPORT void FlutterDesktopViewResizeView(FlutterDesktopViewRef view,
+                                                 int32_t width,
+                                                 int32_t height);
```
These are still under development. Could you please add appropriate comments to prevent these APIs from being used in commercial applications?

JSUYA replied:
I added description.Please review again if desc is appropriate

swift-kim commented on shell/platform/tizen/channels/window_channel.cc:
```diff
@@ -57,7 +57,7 @@ void WindowChannel::HandleMethodCall(
 
-    TizenWindow::Geometry geometry = window_->GetWindowGeometry();
+    TizenViewBase::Geometry geometry = window_->GetRenderTargetGeometry();
     // FIXME: Use SetWindowGeometry() instead of OnGeometryChanged()
     // After the SetWindowGeometry was successfully executed, I expected a
```
`SetWindowGeometry` has been renamed to `SetRenderTargetGeometry` thus this comment needs an update.

swift-kim commented on shell/platform/tizen/tizen_view_elementary.cc:
```diff
@@ -0,0 +1,374 @@
+  evas_object_move(image_, geometry.left, geometry.top);
+}
+
+TizenViewBase::Geometry TizenViewElementary::GetScreenGeometry() {
+  Geometry result;
+  evas_object_geometry_get(image_, &result.left, &result.top, &result.width,
+                           &result.height);
+  return result;
```
The implementation doesn't actually match the method name (the method returns the image's geometry but not the screen geometry). I think `GetScreenGeometry` can be moved to the `TizenWindow` base class as it's only used within `window_channel.cc`.

swift-kim commented on shell/platform/tizen/tizen_window.h:
```diff
@@ -6,71 +6,31 @@
-  virtual void* GetRenderTarget() = 0;
-
-  virtual void* GetRenderTargetDisplay() = 0;
+  virtual void* GetWindowHandle();
```
Question: What's the difference between `GetWindowId` and `GetWindowHandle`? The both return a pointer (handle) to the underlying window instance but I can't clearly get what they are exactly for from their names. Can we refine this part as well?

JSUYA replied:
I think that both return value is different.

| renderer  | Id | Handle |
| ------------- | ------------- | ------------- |
| Elementary | ecore window | elm window |
| wl2 | ecore wl2 window id | ecore wl2 window |

I changed `ecore_evas_window_get(ecore_evas_ecore_evas_get(evas_object_evas_get(win_)))` to `elm_win_window_id_get(elm_win_);`.
elm_win_window_id_get() has the same behavior if the argument is elm_win.

swift-kim replied:
Thanks for the clarification!

swift-kim commented on shell/platform/tizen/tizen_view_base.h:
```diff
@@ -0,0 +1,78 @@
+#ifndef EMBEDDER_TIZEN_VIEW_BASE_H_
+#define EMBEDDER_TIZEN_VIEW_BASE_H_
+
+#include <cstdint>
```
```suggestion
#include <cstdint>
#include <memory>
```

swift-kim commented on shell/platform/tizen/tizen_view_elementary.cc:
```diff
@@ -0,0 +1,374 @@
+}
+
+TizenViewElementary::~TizenViewElementary() {
+  UnregisterEventHandlers();
```
Shouldn't you destroy anything (such as `image_`)?

swift-kim commented on shell/platform/tizen/tizen_window.h:
```diff
@@ -6,71 +6,31 @@
 #ifndef EMBEDDER_TIZEN_WINDOW_H_
 #define EMBEDDER_TIZEN_WINDOW_H_
 
+#include "flutter/shell/platform/tizen/tizen_view_base.h"
```
This line should come after stdlib header includes.

swift-kim commented on shell/platform/tizen/tizen_window.h:
```diff
@@ -6,71 +6,31 @@
-  TizenInputMethodContext* input_method_context() {
-    return input_method_context_.get();
-  }
+  std::string GetType() override { return "window"; };
```
Can we use a dedicated enum instead of a string to represent the type?

Actually this kind of runtime type check is discouraged by [the style guide](https://google.github.io/styleguide/cppguide.html#Run-Time_Type_Information__RTTI_). We may need to either redesign our class hierarchy, or refactor the code so that the type check becomes unnecessary.

> Do not hand-implement an RTTI-like workaround. The arguments against RTTI apply just as much to workarounds like class hierarchies with type tags.

JSUYA replied:
I modified it to use enum.

swift-kim commented on shell/platform/tizen/tizen_window_elementary.h:
```diff
@@ -16,16 +16,16 @@ namespace flutter {
 class TizenWindowElementary : public TizenWindow {
  public:
-  TizenWindowElementary(Geometry geometry,
+  TizenWindowElementary(TizenViewBase::Geometry geometry,
```
nit:
```suggestion
  TizenWindowElementary(Geometry geometry,
```
This also applies to `TizenWindowEcoreWl2` and `TizenViewElementary`.

JSUYA commented on shell/platform/tizen/channels/window_channel.cc:
```diff
@@ -35,7 +35,7 @@ void WindowChannel::HandleMethodCall(
 
   if (method_name == "getWindowGeometry") {
-    TizenWindow::Geometry geometry = window_->GetWindowGeometry();
+    TizenViewBase::Geometry geometry = window_->GetRenderTargetGeometry();
```
I changed TizenViewBase::Geometry to TizenGeometry.

JSUYA commented on shell/platform/tizen/flutter_tizen_ecore.cc:
```diff
@@ -52,3 +53,25 @@ FlutterDesktopViewRef FlutterDesktopViewCreateFromNewWindow(
+    FlutterDesktopEngineRef engine,
+    void* parent) {
+  // FIXME: Ecore WL2 renderer does not support this yet.
+  FT_UNIMPLEMENTED();
```
I deleted wrong comment.

JSUYA commented on shell/platform/tizen/public/flutter_tizen.h:
```diff
@@ -135,6 +143,24 @@ FLUTTER_EXPORT FlutterDesktopViewRef FlutterDesktopViewCreateFromNewWindow(
     FlutterDesktopEngineRef engine);
 
+// Creates a view that hosts and displays the given engine instance.
+// (@warning: This API is improving. There is no guarantee that exact result.)
```
I modified description. Please check again?
If it's still wrong, please suggest a suitable sentence.
```
// @warning This API is a work-in-progress and may change.
```
This sentence was referenced from the android code.

JSUYA commented on shell/platform/tizen/tizen_view_elementary.h:
```diff
@@ -0,0 +1,71 @@
+
+  void* GetRenderTarget() override { return container_; }
+
+  void* GetRenderTargetDisplay() override { return image_; }
```
I modified `RenderTarget` return a image as per your review.
and I don't think we need `GetRenderTargetDisplay()` in `tizen_view_xxx` . So I moved `GetRenderTargetDisplay()` to `TizenWindow`. 
I added `GetRenderTargetContainer()` in `TizenView` but I'm not sure if this is appropriate.
What do you think? Is the name appropriate?

swift-kim replied:
The name looks good, but it seems `TizenWindowElementary::GetRenderTargetDisplay` is still returning a wrong object.

JSUYA replied:
I changed TizenWindowElementary::GetRenderTargetDisplay to return nullptr, it doesn't necessarily return elm_win_.

JSUYA commented on shell/platform/tizen/tizen_view.h:
```diff
@@ -0,0 +1,33 @@
+// Copyright 2022 Samsung Electronics Co., Ltd. All rights reserved.
+// Copyright 2013 The Flutter Authors. All rights reserved.
```
I copied this code from `TizenWindow`. As far as I know `TizenWindow` is a reference from Flutter-window.
Then, can I remove also code in tizen_view_base.h in the same reason?

swift-kim replied:
I think you can, since the code has been written by us.

JSUYA commented on shell/platform/tizen/tizen_view_elementary.cc:
```diff
@@ -0,0 +1,374 @@
+          if (self->event_layer_ == object) {
+            auto* mouse_event =
+                reinterpret_cast<Evas_Event_Mouse_Move*>(event_info);
+            mouse_event->event_flags = static_cast<Evas_Event_Flags>(
+                mouse_event->event_flags | EVAS_EVENT_FLAG_ON_HOLD);
```
This code is the way to that key events are no longer dispatched to other objects.

swift-kim replied:
The reason why I set this comment as "unresolved" was that the suggested change had not been applied. Yes, the suggestion is kind of a matter of preference and whether to apply the change up to your decision, but the issue is technically not resolved.

JSUYA replied:
This is my mistake. (Other codes in this file have been modified). Thank you for check.

JSUYA replied:
(Commit suggestions automatically switch to resolved)

JSUYA commented on shell/platform/tizen/tizen_view_elementary.cc:
```diff
@@ -0,0 +1,374 @@
+  elm_object_style_set(event_layer_, "transparent");
+  evas_object_color_set(event_layer_, 0, 0, 0, 0);
+  EvasObjectResize(event_layer_, initial_geometry_.width,
+                   initial_geometry_.height);
```
I modified some code.
We can "match parent" using evas_object_size_hint_weight_set().
but, some elementary widgets require the size of a child.

JSUYA commented on shell/platform/tizen/tizen_view_elementary.cc:
```diff
@@ -0,0 +1,374 @@
+  // Views do not need to have window info.
+  // However, it is necessary to bind a special key for each profile(in this
+  // case). This part is modified after refactoring related to the key event
+  // code.
```
I removed comment.

JSUYA commented on shell/platform/tizen/tizen_view_base.h:
```diff
@@ -0,0 +1,78 @@
+  }
+
+ protected:
+  FlutterTizenView* view_ = nullptr;
```
I agree about the interdependence issue between `FlutterTizenView` and `TizenViewBase`.
`FlutterTizenEngine` also refer to `FlutterTizenView`. Therefore, I think that we can add the interface of `FlutterTizenView` (`FlutterTizenViewDelegate`?).  And make `FlutterTizenEngine` and `TizenViewBase` refer to `FlutterTizenViewDelegate(?)`.

I will consider it after this PR is applied.

JSUYA commented on shell/platform/tizen/tizen_window_ecore_wl2.cc:
```diff
@@ -381,14 +381,14 @@ void TizenWindowEcoreWl2::DestroyWindow() {
 }
 
-void TizenWindowEcoreWl2::SetWindowGeometry(Geometry geometry) {
+void TizenWindowEcoreWl2::SetRenderTargetGeometry(TizenGeometry geometry) {
```
`TizenViewElementary::SetRenderTargetGeometry` resizes `image_`(min, max_set) and `container_` will changed with it.
We can explicitly resize(min, max_set) `container_`(and `event_layer_`) as well, but I changed the code to not do that.

I think the problem is the method name. I changed this to `SetGeometry` / `GetGeometry`. (Also changed `ResizeRenderTargetWithRotation` -> `ResizeithRotation`).

swift-kim commented on shell/platform/tizen/tizen_view.h:
```diff
@@ -0,0 +1,35 @@
+#ifndef EMBEDDER_TIZEN_VIEW_H_
+#define EMBEDDER_TIZEN_VIEW_H_
+
+#include <string>
```
Unused.

swift-kim commented on shell/platform/tizen/tizen_view.h:
```diff
@@ -0,0 +1,35 @@
+  TizenViewType GetType() override { return TizenViewType::kView; };
+
+ protected:
+  explicit TizenView(int width, int height)
```
You may prefer `int32_t` over `int`.

swift-kim commented on shell/platform/tizen/tizen_view_elementary.cc:
```diff
@@ -0,0 +1,353 @@
+  };
+  evas_object_event_callback_add(event_layer_, EVAS_CALLBACK_MOUSE_UP,
+                                 evas_object_callbacks_[EVAS_CALLBACK_MOUSE_UP],
+                                 this);
+  evas_object_callbacks_[EVAS_CALLBACK_MOUSE_MOVE] =
```
nit: missing newline between the two lines.

swift-kim commented on shell/platform/tizen/tizen_view_elementary.h:
```diff
@@ -0,0 +1,69 @@
+#ifndef EMBEDDER_TIZEN_VIEW_ELEMENTARY_H_
+#define EMBEDDER_TIZEN_VIEW_ELEMENTARY_H_
+
+#include "flutter/shell/platform/tizen/tizen_view.h"
+
+#define EFL_BETA_API_SUPPORT
+#include <Ecore.h>
+#include <Elementary.h>
+
+#include <unordered_map>
```
```suggestion
#define EFL_BETA_API_SUPPORT
#include <Ecore.h>
#include <Elementary.h>

#include <unordered_map>

#include "flutter/shell/platform/tizen/tizen_view.h"
```

Also applies to `tizen_window_*.h`.

JSUYA replied:
I modified the code according to your advice and style guide.
Could you please check again that my understanding is correct?

bbrto21 commented on shell/platform/tizen/flutter_tizen_elementary.cc:
```diff
@@ -52,3 +50,48 @@ FlutterDesktopViewRef FlutterDesktopViewCreateFromNewWindow(
+                              int32_t height) {
+  auto* view = reinterpret_cast<flutter::FlutterTizenView*>(view_ref);
+  flutter::TizenGeometry view_geometry = {0, 0, width, height};
+  view->tizen_view()->OnGeometryChanged(view_geometry);
```
In my opinion, the embedder APIs related to View should be something like C wrapper of FlutterTizenView interface.
so, I believe that to provide the resizing, it has to work internally through FlutterTizenView's interface.

This is the basic design I was initially thinking of, so it's open for discussion and you don't need to fix it right away. I'm curious about your opinion.

swift-kim replied:
I don't agree. `FlutterTizenView` is simply an abstraction layer for a Flutter view and it should not be considered as a wrapper of `TizenView`.

bbrto21 replied:
> `FlutterTizenView` is simply an abstraction layer for a Flutter view and it should not be considered as a wrapper of `TizenView`.

What you said is a simple principle I came up with when I first designed it. 
When I first designed FlutterDeskopViewXXX APIs, I had the following design in mind.
```
void FlutterDesktopViewXXX()
{
   ...
    view->XXX()
   ...
}
```

Currently, FlutterTizenView may have different detailed behaviors depending on TizenViewType. 
So I think this idea may not be suitable. That's why I was curious about your thoughts :)

JSUYA replied:
I agree to reduce the abstraction level (depth) of required `FlutterDesktopViewXXX`.
In the case of `resize` rather than simple property change, I think it is better to handle it with `view->Rsize()`.
If the ViewType check is handled inside the FlutterTizenView, I think it is easy to fix when the type check is no longer needed through refactoring.

bbrto21 commented on shell/platform/tizen/flutter_tizen_view.cc:
```diff
@@ -40,10 +41,10 @@ const std::vector<std::string> kBindableSystemKeys = {
+FlutterTizenView::FlutterTizenView(std::unique_ptr<TizenViewBase> tizen_view)
+    : tizen_view_(std::move(tizen_view)) {
+  tizen_view_->SetView(this);
+  tizen_view_->BindKeys(kBindableSystemKeys);
```
When the type of `tizen_view_` is not window, I don't think this is the role of the Embedder anymore. I think it should be fixed later.

swift-kim replied:
Yes, that makes sense. I think we can move `BindKeys` from `TizenViewBase` to `TizenWindow`.

JSUYA replied:
I think the view also needs a bind to a special key.
We may also need the TV remote's back key or other special key inside the View.
Since key binding is platform dependent, I think that the information provided for VD profile can be same with either Window or View.

This has already been talked about in private chats.
If necessary, I hope this is discussed in a separate PR regarding special key bindings.

bbrto21 replied:
Yes, we can handle it in another PR.

swift-kim replied:
> I think the view also needs a bind to a special key.
We may also need the TV remote's back key or other special key inside the View.
Since key binding is platform dependent, I think that the information provided for VD profile can be same with either Window or View.

Sorry. I don't still get why it should be the embedder's role to register the special keys. Isn't is possible for the parent app to call `eext_win_keygrab_set` with its own window as necessary?

JSUYA replied:
> Sorry. I don't still get why it should be the embedder's role to register the special keys. Isn't is possible for the parent app to call eext_win_keygrab_set with its own window as necessary?

Using eext_win_keygrab_set, the app user can also bind a key to the app window.

my opinion is that the flutter app(window and view) created with flutter-tizen can provide the same key bind within the VD profile. If the user has experienced that BackKey works in flutter-app (win), but does not add it to the app in the view, it may not work. Of course, not binding to the app's window may mean that a special key is not used.
This is just my opinion.

however, it seems problematic for the view to modify the properties of the window.
I can also delete the view's bind code from this PR. Should I do that now?
@bbrto21 told me on chat that he would proceed with the modify related to key bind.

bbrto21 commented on shell/platform/tizen/tizen_view_elementary.cc:
```diff
@@ -0,0 +1,354 @@
+    initial_height_ = parent_height;
+  }
+
+  container_ = elm_table_add(parent_);
```
Could you please leave comments somewhere why we need a container in the form of a table?

JSUYA replied:
The elm_table is one of the ways that can be used at C API level when overlapping widgets in elementary.
I don't think this is a special thing enough to reason leaving a comment.

bbrto21 replied:
The container can eventually reach the user via the `FlutterDesktopViewGetEvasObject` under current implementations. 
Tables typically provide complex functionality for their unique layout flow. so it's generally considered to be ineffective when simply it occupying one box area, although not fatal. therefore, the reason for choosing a table among the various widgets should be reasonable.

Also, I think we have a duty to inform the user what the object is essentially returned through the `FlutterDesktopViewGetEvasObject`.

JSUYA replied:
I don't think that.
The API name and Description describe that XXXGetEvasObject returns Evas_Object.
I do not want to express that the returned object has the properties of Table, and I do not want the user to use the elm_table API.
The user should only be able to use the properties that Evas_Object can have.

bbrto21 reviewed (approved):
I think there are a few things to discuss, but it seems to be solvable in further PR.
and it seems to work just as the creator intended. LGTM.

swift-kim commented:
Please don't mark my comment as "resolved" by yourself. I'll do that when necessary.

JSUYA commented:
> Please don't mark my comment as "resolved" by yourself. I'll do that when necessary.
> 
> If I checked your code and didn't close the corresponding comment, there's always a reason for it.

um... I think of `Resolved` as a simple expression of the comment("Done") that your review has been accepted.
If you have a reason just do `Unresolved` again.
If it is the rule of flutter-tizen that the reviewer checks `Resolved`, I will follow it.

swift-kim commented:
I told because you were keeping closing comments that I had set as "unresolved". Please leave a simple comment such as "Done" if the issue has been resolved.

bbrto21 commented:
@swift-kim Do you have anything more to discuss?

bbrto21 commented:
> I told because you were keeping closing comments that I had set as "unresolved". Please leave a simple comment such as "Done" if the issue has been resolved.

Is this a common practice? for those who have just joined the project this can be an inconvenience.

swift-kim commented:
@bbrto21 No, my comment was specifically for him. It's not a rule that applies to everyone.

> Do you have anything more to discuss?

It seems there are still some unresolved issues that have not been answered by the author.

- https://github.com/flutter-tizen/engine/pull/292#discussion_r895389422
- https://github.com/flutter-tizen/engine/pull/292#discussion_r895402744
- https://github.com/flutter-tizen/engine/pull/292#discussion_r892013696

swift-kim reviewed (approved):
Thank you!
